File: packages/core/src/__tests__/join.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { Engine } from '../engine.js'

describe('Join patterns (on([...]))', () => {
  let engine: Engine

  beforeEach(() => {
    engine = new Engine()
  })

  it('should fire join only when all inputs are ready', () => {
    const a = engine.event<number>('a')
    const b = engine.event<string>('b')
    const results: string[] = []

    engine.on([a, b], (numVal: number, strVal: string) => {
      results.push(`${numVal}-${strVal}`)
    })

    // Only emit a — join should not fire
    engine.emit(a, 42)
    expect(results).toEqual([])

    // Now emit b — join should fire
    engine.emit(b, 'hello')
    expect(results).toEqual(['42-hello'])
  })

  it('should consume one event from each input type per join', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const results: number[] = []

    engine.on([a, b], (av: number, bv: number) => {
      results.push(av + bv)
    })

    engine.emit(a, 1)
    engine.emit(a, 2)
    engine.emit(b, 10)
    // Should use first a (1) + b (10)
    expect(results).toContain(11)
  })

  it('should handle join in a diamond DAG via nested emits', () => {
    const src = engine.event<number>('src')
    const left = engine.event<number>('left')
    const right = engine.event<number>('right')
    const results: number[] = []

    engine.on(src, (x) => engine.emit(left, x * 2))
    engine.on(src, (x) => engine.emit(right, x * 3))
    engine.on([left, right], (l: number, r: number) => {
      results.push(l + r)
    })

    engine.emit(src, 10)
    expect(results).toEqual([50]) // 10*2 + 10*3
  })

  it('should handle join with cascading output via emit', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const joined = engine.event<number>('joined')
    const final = engine.event<string>('final')
    const results: string[] = []

    engine.on([a, b], (av: number, bv: number) => {
      engine.emit(joined, av + bv)
    })
    engine.on(joined, (v) => {
      engine.emit(final, `result:${v}`)
    })
    engine.on(final, (v) => results.push(v))

    engine.emit(a, 3)
    engine.emit(b, 7)
    expect(results).toEqual(['result:10'])
  })

  it('should not fire join when only some inputs are available', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const c = engine.event<number>('c')
    const results: number[] = []

    engine.on([a, b, c], (av: number, bv: number, cv: number) => {
      results.push(av + bv + cv)
    })

    engine.emit(a, 1)
    engine.emit(b, 2)
    expect(results).toEqual([])

    engine.emit(c, 3)
    expect(results).toEqual([6])
  })

  it('should handle multiple join firings as events accumulate', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const results: number[] = []

    engine.on([a, b], (av: number, bv: number) => {
      results.push(av + bv)
    })

    engine.emit(a, 1)
    engine.emit(b, 10)
    expect(results).toEqual([11])

    engine.emit(a, 2)
    engine.emit(b, 20)
    expect(results).toEqual([11, 22])
  })

  it('should unsubscribe join handler', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const results: number[] = []

    const unsub = engine.on([a, b], (av: number, bv: number) => {
      results.push(av + bv)
    })

    engine.emit(a, 1)
    engine.emit(b, 10)
    expect(results).toEqual([11])

    unsub()

    engine.emit(a, 2)
    engine.emit(b, 20)
    // Should not fire again
    expect(results).toEqual([11])
  })
})

File: packages/dart/test/join_test.dart
import 'package:test/test.dart';
import 'package:pulse/pulse.dart';

void main() {
  late Engine engine;

  setUp(() {
    engine = Engine();
  });

  test('should fire join only when all inputs are ready', () {
    final a = engine.event<int>('a');
    final b = engine.event<String>('b');
    final out = engine.event<String>('out');
    final results = <String>[];

    engine.join<String>(
      inputs: [a, b],
      output: out,
      action: (payloads) => '${payloads[0]}-${payloads[1]}',
    );
    engine.on<String>(out, (v) => results.add(v));

    // Only emit a -- join should not fire.
    engine.emit(a, 42);
    expect(results, equals([]));

    // Now emit b -- join should fire.
    engine.emit(b, 'hello');
    expect(results, equals(['42-hello']));
  });

  test('should consume one event from each input type per join', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final out = engine.event<int>('out');
    final results = <int>[];

    engine.join<int>(
      inputs: [a, b],
      output: out,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.on<int>(out, (v) => results.add(v));

    engine.emit(a, 1);
    engine.emit(a, 2);
    engine.emit(b, 10);
    // Should use first a (1) + b (10).
    expect(results, contains(11));
  });

  test('should support guard on join -- passing guard', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final out = engine.event<int>('out');
    final results = <int>[];

    engine.join<int>(
      inputs: [a, b],
      output: out,
      guard: (payloads) => (payloads[0] as int) + (payloads[1] as int) > 10,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.on<int>(out, (v) => results.add(v));

    engine.emit(a, 8);
    engine.emit(b, 5);
    expect(results, equals([13]));
  });

  test('should support guard on join -- failing guard blocks join', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final out = engine.event<int>('out');
    final results = <int>[];

    engine.join<int>(
      inputs: [a, b],
      output: out,
      guard: (payloads) => (payloads[0] as int) + (payloads[1] as int) > 100,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.on<int>(out, (v) => results.add(v));

    engine.emit(a, 2);
    engine.emit(b, 3);
    // Guard fails: 2+3=5, not > 100.
    expect(results, equals([]));
  });

  test('should handle join in a diamond DAG', () {
    final src = engine.event<int>('src');
    final left = engine.event<int>('left');
    final right = engine.event<int>('right');
    final merged = engine.event<int>('merged');
    final results = <int>[];

    engine.pipe<int, int>(src, left, (x) => x * 2);
    engine.pipe<int, int>(src, right, (x) => x * 3);
    engine.join<int>(
      inputs: [left, right],
      output: merged,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.on<int>(merged, (v) => results.add(v));

    engine.emit(src, 10);
    expect(results, equals([50])); // 10*2 + 10*3
  });

  test('should handle join with cascading output', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final joined = engine.event<int>('joined');
    final finalEv = engine.event<String>('final');
    final results = <String>[];

    engine.join<int>(
      inputs: [a, b],
      output: joined,
      action: (payloads) => (payloads[0] as int) + (payloads[1] as int),
    );
    engine.pipe<int, String>(joined, finalEv, (v) => 'result:$v');
    engine.on<String>(finalEv, (v) => results.add(v));

    engine.emit(a, 3);
    engine.emit(b, 7);
    expect(results, equals(['result:10']));
  });

  test('should not fire join when only some inputs are available', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final out = engine.event<int>('out');
    final results = <int>[];

    engine.join<int>(
      inputs: [a, b, c],
      output: out,
      action: (payloads) =>
          (payloads[0] as int) + (payloads[1] as int) + (payloads[2] as int),
    );
    engine.on<int>(out, (v) => results.add(v));

    engine.emit(a, 1);
    engine.emit(b, 2);
    expect(results, equals([]));

    engine.emit(c, 3);
    expect(results, equals([6]));
  });

  test('should work with onAll for side-effect-only joins', () {
    final a = engine.event<int>('a');
    final b = engine.event<String>('b');
    final results = <String>[];

    engine.onAll([a, b], (payloads) {
      results.add('${payloads[0]}-${payloads[1]}');
    });

    engine.emit(a, 42);
    expect(results, isEmpty);

    engine.emit(b, 'hello');
    expect(results, equals(['42-hello']));
  });
}
